opendir.c

```diff
@@ -47,23 +47,91 @@ struct watchman_dir_handle {
   struct watchman_dir_ent ent;
 };
 
+/* Opens a directory, strictly prohibiting opening any symlinks
+ * in any component of the path on systems that make this
+ * reasonable and easy to do.
+ * We do this by opening / and then walking down the tree one
+ * directory component at a time.
+ */
+static int open_dir_as_fd_no_symlinks(const char *path, int flags) {
+#ifdef HAVE_OPENAT
+  char *pathcopy;
+  int fd = -1;
+  int err;
+  char *elem, *next;
+
+  if (path[0] != '/') {
+    errno = EINVAL;
+    return -1;
+  }
+
+  pathcopy = strdup(path);
+  if (!pathcopy) {
+    return -1;
+  }
+
+  fd = open("/", O_NOFOLLOW | O_CLOEXEC);
+  if (fd == -1) {
+    err = errno;
+    free(pathcopy);
+    errno = err;
+    return -1;
+  }
+
+  elem = pathcopy + 1;
+  while (elem) {
+    int next_fd;
+
+    next = strchr(elem, '/');
+    if (next) {
+      *next = 0;
+      next++;
+    }
+
+    // Ensure that we're CLOEXEC on all intermediate components of
+    // path, but use the flags the caller passed to us when we reach
+    // the leaf of the tree
+    next_fd = openat(fd, elem, (next ? O_CLOEXEC : flags)|O_NOFOLLOW);
+    if (next_fd == -1) {
+      err = errno;
+      close(fd);
+      free(pathcopy);
+      errno = err;
+      return -1;
+    }
+
+    // walk down to next level
+    elem = next;
+    close(fd);
+    fd = next_fd;
+  }
+
+  // This was the final component in the path
+  free(pathcopy);
+  return fd;
+#else
+  return open(path, flags);
+#endif
+}
+
 /* Opens a directory making sure it's not a symlink */
 DIR *opendir_nofollow(const char *path)
 {
 #ifdef _WIN32
   return win_opendir(path, 1 /* no follow */);
 #else
-  int fd = open(path, O_NOFOLLOW | O_CLOEXEC);
+  int fd = open_dir_as_fd_no_symlinks(path, O_NOFOLLOW | O_CLOEXEC);
   if (fd == -1) {
     return NULL;
   }
-#if defined(__APPLE__)
+# if !defined(HAVE_FDOPENDIR) || defined(__APPLE__)
+  /* fdopendir doesn't work on OS X */
```

Comment:
In earlier versions, we had a configure check for it, it would link, but it would always error out :-/
However, we now use the bulk stat stuff since 10.10, so we avoid this issue in most systems on OS X